### PR TITLE
Fix false positive tests in CI for Windows

### PR DESCRIPTION
**Pull request by UebelAndre**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently the integration test runner does not report errors for windows.

Issue Number: N/A

## What is the new behavior?
The test runner has been fixed and the examples/tests updated to be compatible with windows.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
This PR includes the following which are both required to get passing windows builds.
- https://github.com/bazelbuild/rules_python/pull/597
- https://github.com/bazelbuild/rules_python/pull/595

### CODE REVIEW

File: version.bzl [hunk 1]
```diff
@@ -18,7 +18,7 @@
 # This version should be updated together with the version of the Bazel
 # in .bazelversion.
 # TODO(alexeagle): assert this is the case in a test
-BAZEL_VERSION = "4.0.0"
+BAZEL_VERSION = "4.1.0"
```
Comment (UebelAndre) [inline version.bzl:21]: This is needed to be able to use `@platforms//:incompatible` to disable the  `diff_test` in `examples/pip_install`. I'm happy to revert this if there's a way to get it working, but I couldn't figure it out :(

**Review by UebelAndre — COMMENTED** [review]
@jheaff1 would you mind taking a look to see if I missed anything? And maybe help me address some TODOs? 😅 🙏

File: examples/pip_install/BUILD [hunk 2]
```diff
@@ -116,4 +106,9 @@ diff_test(
     name = "test_query_result",
     file1 = "expected",
     file2 = "yamllint_lib_by_version",
+    # TODO: It's unclear as to why this does not work on windows
```
[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/BUILD:109]: @UebelAndre  Sorry for the late reply. I checked out your branch and ran this diff_test, which passes for me. I'm using Python 3.10, maybe update the Python version used in CI?

[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/BUILD:109]: Passes with bazel 4.0.0 also

File: examples/pip_install/pip_install_test.py [hunk 3]
```diff
@@ -10,50 +11,37 @@ class PipInstallTest(unittest.TestCase):
...
-        self.assertEqual(proc.stdout.decode("utf-8").strip(), "yamllint 1.26.3")
-
-        # yamllint entry_point is of the form `def run(argv=None):`
-        with self.assertRaises(subprocess.CalledProcessError) as context:
-            subprocess.run(
-                [entry_point, "--option-does-not-exist"],
-                check=True,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
+        # TODO: It's unclear as to why `$(rootpath :yamllint)` does not
```
[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/pip_install_test.py:15]: In my projects I always use the [`--nobuild_python_zip`](https://docs.bazel.build/versions/main/command-line-reference.html#flag--build_python_zip) argument to bazel, as I find that it adds a few seconds to each `bazel run` of a `py_binary`. If i'm not mistaken, the reason that `--build_python_zip` is on by default on Windows only is that typically Windows does not support symlinks, however I enabled symlinks on my Windows PC.

With `--nobuild_python_zip`, the `pip_install_test` passes without your changes.

With or without the `--nobuild_python_zip` flag, `rootpath` is expanded to `C:\msys64\home\jheaf\_bazel_jheaf\6paoe6i2\execroot\example_repo\bazel-out\x64_windows-fastbuild\bin\pip_install_test.exe.runfiles\example_repo`.

Yet the path of the executable when `--nobuild_python_zip` is not set is: C:\Users\jheaf\AppData\Local\Temp\Bazel.runfiles__klds23u\runfiles\example_repo.

Perhaps it is a bug in bazel.

Could the CI config of rules_python be changed to use the `--nobuild_python_zip` flag? It already uses `--enable_runfiles`, which requires symlinks to be enabled on Windows.

File: examples/pip_install/BUILD (hunk 2 above)
[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/BUILD:109]: Ah so the test passes if I don't use the flag `--enable_runfiles`, which is set in the CI build of this repo, I believe.

[the remark this answers is not in the document]
Comment (UebelAndre) [reply examples/pip_install/BUILD:109]: Trying this now.

[the remark this answers is not in the document]
Comment (UebelAndre) [reply examples/pip_install/BUILD]: It works! Must be an issue on my machine :(

File: examples/pip_install/pip_install_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (UebelAndre) [reply examples/pip_install/pip_install_test.py:15]: I can try adding the flags but do you know if there's a way to get this to work in a cleaner way without them?

Would https://github.com/bazelbuild/bazel/blob/4.0.0/tools/python/runfiles/runfiles.py help?

File: examples/pip_install/BUILD (hunk 2 above)
[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/BUILD]: It seems to be an issue with skylib's `diff_test` rule when using `--enable_runfiles` on windows. This was resolved in https://github.com/bazelbuild/bazel-skylib/commit/43a545a67ea2b9673365745b1b5372170d730922. Simply using the latest release of bazel-skylib in the `examples/pip_install` workspace resolved the issue

**Comment by jheaff1** [comment]
I’m suprised the diff_tests works in CI, as diff_test is broken when using `—runfiles` until skylib v1.1.1 

https://github.com/bazelbuild/bazel-skylib/commit/43a545a67ea2b9673365745b1b5372170d730922

File: examples/pip_install/pip_install_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jheaff1) [reply examples/pip_install/pip_install_test.py:15]: Hmm perhaps that python script could indeed help, I’ve never seen it before

[the remark this answers is not in the document]
Comment (UebelAndre) [reply examples/pip_install/pip_install_test.py]: It does! I'm gonna squash the changes but I think this works out pretty well 😄

**Comment by UebelAndre** [comment]
@jheaff1 Thanks so much for your help!

**Comment by jheaff1** [comment]
> @jheaff1 Thanks so much for your help!

My pleasure, although you did 99% of the work, so thank you! 😂

**Comment by UebelAndre** [comment]
As all other changes have been merged, this PR is just left with https://github.com/bazelbuild/rules_python/pull/597

**Review by f0rmiga — APPROVED** [review]

**Pull request merged**